1. Para iniciar, importamos Pandas

In [ ]:
import pandas as pd

2. Comenzamos el Proceso de ETL. Cargamos los 2 archivos .csv (sales_data y customer_data).

In [ ]:
customer_df = pd.read_csv('customer_data.csv')
sales_df = pd.read_csv('sales_data.csv')

**Conclusión:** Con las sentencias realizadas, a través de Pandas se cargaron y prepararon los datos para poder visualizarlos a través de dos DataFrames ("customer_df" y "sales_df").

3. Solicitamos mostrar los datos de cada Data Frame.

In [ ]:
customer_df

customer_id  gender   age payment_method
0         C241288  Female  28.0    Credit Card
1         C111565    Male  21.0     Debit Card
2         C266599    Male  20.0           Cash
3         C988172  Female  66.0    Credit Card
4         C189076  Female  53.0           Cash
...           ...     ...   ...            ...
99452     C441542  Female  45.0    Credit Card
99453     C569580    Male  27.0           Cash
99454     C103292    Male  63.0     Debit Card
99455     C800631    Male  56.0           Cash
99456     C273973  Female  36.0    Credit Card

[99457 rows x 4 columns]

**Conclusión:** Al solicitar ver el Data Frame customer_df, además de poder visualizar los datos, obtenemos referencias importantes, como la cantidad de filas (99457) y la cantidad de columnas (4).

In [ ]:
sales_df

invoice_no customer_id         category  quantity    price invoice_date  \
0        I138884     C241288         Clothing         5  1500.40   05-08-2022   
1        I317333     C111565            Shoes         3  1800.51   12-12-2021   
2        I127801     C266599         Clothing         1   300.08   09-11-2021   
3        I173702     C988172            Shoes         5  3000.85   16-05-2021   
4        I337046     C189076            Books         4    60.60   24-10-2021   
...          ...         ...              ...       ...      ...          ...   
99452    I219422     C441542         Souvenir         5    58.65   21-09-2022   
99453    I325143     C569580  Food & Beverage         2    10.46   22-09-2021   
99454    I824010     C103292  Food & Beverage         2    10.46   28-03-2021   
99455    I702964     C800631       Technology         4  4200.00   16-03-2021   
99456    I232867     C273973         Souvenir         3    35.19   15-10-2022   

          shopping_mall  
0                Kanyon  
1        Forum Istanbul  
2             Metrocity  
3          Metropol AVM  
4                Kanyon  
...                 ...  
99452            Kanyon  
99453    Forum Istanbul  
99454         Metrocity  
99455      Istinye Park  
99456  Mall of Istanbul  

[99457 rows x 7 columns]

**Conclusión:** Al igual que con el Data Frame customer_df, podemos visualizar los datos, la cantidad de filas (99457) y la cantidad de columnas (7) que tiene el Data Frame sales_df.

Utilizamos describe() para conocer más sobre los DataFrame

In [ ]:
# Describir los DataFrames
print("Descripción del DF Sales")
print(sales_df.describe())
print("\nDescripción del DF Customer")
print(customer_df.describe())


Descripción del DF Sales
           quantity         price
count  99457.000000  99457.000000
mean       3.003429    689.256321
std        1.413025    941.184567
min        1.000000      5.230000
25%        2.000000     45.450000
50%        3.000000    203.300000
75%        4.000000   1200.320000
max        5.000000   5250.000000

Descripción del DF Customer
                age
count  99338.000000
mean      43.425859
std       14.989400
min       18.000000
25%       30.000000
50%       43.000000
75%       56.000000
max       69.000000


También podemos describir cada DataFrame para obtener estadísticas básicas

Verificamos si existen valores nulos en los DataFrames

In [ ]:
# Verificar valores nulos en el DataFrame
valores_nulos_sales = sales_df.isnull().sum()
print(valores_nulos_sales)


invoice_no       0
customer_id      0
category         0
quantity         0
price            0
invoice_date     0
shopping_mall    0
dtype: int64


In [ ]:
# Verificar valores nulos en el DataFrame
valores_nulos_customer = customer_df.isnull().sum()
print(valores_nulos_customer)


customer_id         0
gender              0
age               119
payment_method      0
dtype: int64


4. Para no tener que trabajar con 2 archivos separados, los unificamos en un solo DataFrame para una mejor visualización. La combinación la realizamos tomando el campo en común "id del cliente".

In [ ]:
unificado_df = pd.merge(sales_df, customer_df, on='customer_id', how='inner')

5. Corroboramos la fusión de los 2 .csv solicitando ver más información sobre las columnas y los tipos de datos que contiene el Data Frame.

In [ ]:
unificado_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99457 entries, 0 to 99456
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   invoice_no      99457 non-null  object 
 1   customer_id     99457 non-null  object 
 2   category        99457 non-null  object 
 3   quantity        99457 non-null  int64  
 4   price           99457 non-null  float64
 5   invoice_date    99457 non-null  object 
 6   shopping_mall   99457 non-null  object 
 7   gender          99457 non-null  object 
 8   age             99338 non-null  float64
 9   payment_method  99457 non-null  object 
dtypes: float64(2), int64(1), object(7)
memory usage: 7.6+ MB


**Conclusión:** Podemos observar que la cantidad de filas se mantiene en el mismo número (99.457), pero se incrementa el número de columnas a 10, ya que contiene las 7 columnas de sales_data.csv y agrega solo 3 de customer_data.csv, ya que la columna ID es la misma en ambos Data Frames / archivos.csv.
A su vez, conocemos los tipos de datos que contiene cada columna (float, int64, object).

6. Visualizamos el nuevo Data Frame.

In [ ]:
unificado_df

invoice_no customer_id         category  quantity    price invoice_date  \
0        I138884     C241288         Clothing         5  1500.40   05-08-2022   
1        I317333     C111565            Shoes         3  1800.51   12-12-2021   
2        I127801     C266599         Clothing         1   300.08   09-11-2021   
3        I173702     C988172            Shoes         5  3000.85   16-05-2021   
4        I337046     C189076            Books         4    60.60   24-10-2021   
...          ...         ...              ...       ...      ...          ...   
99452    I219422     C441542         Souvenir         5    58.65   21-09-2022   
99453    I325143     C569580  Food & Beverage         2    10.46   22-09-2021   
99454    I824010     C103292  Food & Beverage         2    10.46   28-03-2021   
99455    I702964     C800631       Technology         4  4200.00   16-03-2021   
99456    I232867     C273973         Souvenir         3    35.19   15-10-2022   

          shopping_mall  gender   age payment_method  
0                Kanyon  Female  28.0    Credit Card  
1        Forum Istanbul    Male  21.0     Debit Card  
2             Metrocity    Male  20.0           Cash  
3          Metropol AVM  Female  66.0    Credit Card  
4                Kanyon  Female  53.0           Cash  
...                 ...     ...   ...            ...  
99452            Kanyon  Female  45.0    Credit Card  
99453    Forum Istanbul    Male  27.0           Cash  
99454         Metrocity    Male  63.0     Debit Card  
99455      Istinye Park    Male  56.0           Cash  
99456  Mall of Istanbul  Female  36.0    Credit Card  

[99457 rows x 10 columns]

**Conclusión:** Al visualizar el nuevo DataFrame observamos los datos de las 10 columnas.

7. Todo el proceso realizado hasta acá, corresponde a la **Extracción de Datos (ETL)** y damos comienzo al proceso de **Transformación**.

Al realizar un análisis previo a la "Transformación", cuando implementamos la visualización de la información de columnas y tipos de datos "unificado_df.info()" notamos que en la columna "age" el número de datos era menor a la cantidad de filas que contiene el DataFrame. En este caso teníamos 2 opciones, recurrir a la eliminación de filas con datos vacíos (que podían afectar a un análisis posterior) o rellenar los valores vacíos/nulos con el promedio, mediana o un valor fijo a través del método ".fillna". Al observar que la diferencia de valores no era tan grande (119 valores de 99.457), optamos eliminar las filas sin datos.

In [ ]:
unificado_df.dropna(inplace=True)
unificado_df

invoice_no customer_id         category  quantity    price invoice_date  \
0        I138884     C241288         Clothing         5  1500.40   05-08-2022   
1        I317333     C111565            Shoes         3  1800.51   12-12-2021   
2        I127801     C266599         Clothing         1   300.08   09-11-2021   
3        I173702     C988172            Shoes         5  3000.85   16-05-2021   
4        I337046     C189076            Books         4    60.60   24-10-2021   
...          ...         ...              ...       ...      ...          ...   
99452    I219422     C441542         Souvenir         5    58.65   21-09-2022   
99453    I325143     C569580  Food & Beverage         2    10.46   22-09-2021   
99454    I824010     C103292  Food & Beverage         2    10.46   28-03-2021   
99455    I702964     C800631       Technology         4  4200.00   16-03-2021   
99456    I232867     C273973         Souvenir         3    35.19   15-10-2022   

          shopping_mall  gender   age payment_method  
0                Kanyon  Female  28.0    Credit Card  
1        Forum Istanbul    Male  21.0     Debit Card  
2             Metrocity    Male  20.0           Cash  
3          Metropol AVM  Female  66.0    Credit Card  
4                Kanyon  Female  53.0           Cash  
...                 ...     ...   ...            ...  
99452            Kanyon  Female  45.0    Credit Card  
99453    Forum Istanbul    Male  27.0           Cash  
99454         Metrocity    Male  63.0     Debit Card  
99455      Istinye Park    Male  56.0           Cash  
99456  Mall of Istanbul  Female  36.0    Credit Card  

[99338 rows x 10 columns]

**Conclusión:** Como pudimos observar al solicitar los datos de las columnas (unificado_df.info()) la columna "age" tenía una menor cantidad de datos que el resto, entonces, al aplicar una limpieza del DataFrame eliminando los valores nulos, nos encontramos que se han eliminado 119 filas, ya que se redujeron de 99.457 a 99.338 filas.

8. Se transforman los datos del formato dd/mm/YYYY al formato YYYY/mm en una nueva columna para obtener un mejor resultado para análisis posteriores.

In [ ]:
unificado_df['invoice_date'] = pd.to_datetime(unificado_df['invoice_date'], format='%d-%m-%Y')
unificado_df['year_month'] = unificado_df['invoice_date'].dt.to_period('M').astype(str)
fecha_global_df = unificado_df['year_month']
unificado_df

invoice_no customer_id         category  quantity    price invoice_date  \
0        I138884     C241288         Clothing         5  1500.40   2022-08-05   
1        I317333     C111565            Shoes         3  1800.51   2021-12-12   
2        I127801     C266599         Clothing         1   300.08   2021-11-09   
3        I173702     C988172            Shoes         5  3000.85   2021-05-16   
4        I337046     C189076            Books         4    60.60   2021-10-24   
...          ...         ...              ...       ...      ...          ...   
99452    I219422     C441542         Souvenir         5    58.65   2022-09-21   
99453    I325143     C569580  Food & Beverage         2    10.46   2021-09-22   
99454    I824010     C103292  Food & Beverage         2    10.46   2021-03-28   
99455    I702964     C800631       Technology         4  4200.00   2021-03-16   
99456    I232867     C273973         Souvenir         3    35.19   2022-10-15   

          shopping_mall  gender   age payment_method year_month  
0                Kanyon  Female  28.0    Credit Card    2022-08  
1        Forum Istanbul    Male  21.0     Debit Card    2021-12  
2             Metrocity    Male  20.0           Cash    2021-11  
3          Metropol AVM  Female  66.0    Credit Card    2021-05  
4                Kanyon  Female  53.0           Cash    2021-10  
...                 ...     ...   ...            ...        ...  
99452            Kanyon  Female  45.0    Credit Card    2022-09  
99453    Forum Istanbul    Male  27.0           Cash    2021-09  
99454         Metrocity    Male  63.0     Debit Card    2021-03  
99455      Istinye Park    Male  56.0           Cash    2021-03  
99456  Mall of Istanbul  Female  36.0    Credit Card    2022-10  

[99338 rows x 11 columns]

**Conclusión:** Se observa con un análisis superficial al visualizar los datos que las fechas (año/mes) coinciden con las originales. A su vez, observamos el incremento en la cantidad de columnas (11).

9. Se procede a borrar la columna “invoice_date” que contenía las fechas originales de los archivos .csv.

In [ ]:
unificado_df.drop('invoice_date', axis=1, inplace=True)
unificado_df

invoice_no customer_id         category  quantity    price  \
0        I138884     C241288         Clothing         5  1500.40   
1        I317333     C111565            Shoes         3  1800.51   
2        I127801     C266599         Clothing         1   300.08   
3        I173702     C988172            Shoes         5  3000.85   
4        I337046     C189076            Books         4    60.60   
...          ...         ...              ...       ...      ...   
99452    I219422     C441542         Souvenir         5    58.65   
99453    I325143     C569580  Food & Beverage         2    10.46   
99454    I824010     C103292  Food & Beverage         2    10.46   
99455    I702964     C800631       Technology         4  4200.00   
99456    I232867     C273973         Souvenir         3    35.19   

          shopping_mall  gender   age payment_method year_month  
0                Kanyon  Female  28.0    Credit Card    2022-08  
1        Forum Istanbul    Male  21.0     Debit Card    2021-12  
2             Metrocity    Male  20.0           Cash    2021-11  
3          Metropol AVM  Female  66.0    Credit Card    2021-05  
4                Kanyon  Female  53.0           Cash    2021-10  
...                 ...     ...   ...            ...        ...  
99452            Kanyon  Female  45.0    Credit Card    2022-09  
99453    Forum Istanbul    Male  27.0           Cash    2021-09  
99454         Metrocity    Male  63.0     Debit Card    2021-03  
99455      Istinye Park    Male  56.0           Cash    2021-03  
99456  Mall of Istanbul  Female  36.0    Credit Card    2022-10  

[99338 rows x 10 columns]

**Conclusión:** Se elimina la columna "invoice_date" y se visualizan los datos para corroborar.

10. Generamos una nueva columna “rango_etario” que contenga la siguiente clasificación: (>24), (25 a 35), (36 a 45), (46 a 60), (<61).

In [ ]:
bins = [0, 24, 35, 45, 60, float('inf')]
labels = ['>24', '25 a 35', '36 a 45', '46 a 60', '<61']
unificado_df['rango_etario'] = pd.cut(unificado_df['age'], bins=bins, labels=labels)
unificado_df

invoice_no customer_id         category  quantity    price  \
0        I138884     C241288         Clothing         5  1500.40   
1        I317333     C111565            Shoes         3  1800.51   
2        I127801     C266599         Clothing         1   300.08   
3        I173702     C988172            Shoes         5  3000.85   
4        I337046     C189076            Books         4    60.60   
...          ...         ...              ...       ...      ...   
99452    I219422     C441542         Souvenir         5    58.65   
99453    I325143     C569580  Food & Beverage         2    10.46   
99454    I824010     C103292  Food & Beverage         2    10.46   
99455    I702964     C800631       Technology         4  4200.00   
99456    I232867     C273973         Souvenir         3    35.19   

          shopping_mall  gender   age payment_method year_month rango_etario  
0                Kanyon  Female  28.0    Credit Card    2022-08      25 a 35  
1        Forum Istanbul    Male  21.0     Debit Card    2021-12          >24  
2             Metrocity    Male  20.0           Cash    2021-11          >24  
3          Metropol AVM  Female  66.0    Credit Card    2021-05          <61  
4                Kanyon  Female  53.0           Cash    2021-10      46 a 60  
...                 ...     ...   ...            ...        ...          ...  
99452            Kanyon  Female  45.0    Credit Card    2022-09      36 a 45  
99453    Forum Istanbul    Male  27.0           Cash    2021-09      25 a 35  
99454         Metrocity    Male  63.0     Debit Card    2021-03          <61  
99455      Istinye Park    Male  56.0           Cash    2021-03      46 a 60  
99456  Mall of Istanbul  Female  36.0    Credit Card    2022-10      36 a 45  

[99338 rows x 11 columns]

**Conclusión:** Para generar una nueva columna (rango_etario) para clasificar a los consumidores acorde a su edad, se implementa la función de Pandas pd.cut(). Primero se definen los límites (bins) y se nombran los rangos (labels). Al finalizar, solicitamos visualizar el DataFrame para un chequeo general. Corroboramos a simple vista que coinciden las edad con los rangos asignados y que l DataFrame se incrementa en una columna más (11).

11. Para dejar más limpio y preparado el DataFrame para el proceso de Carga, se eliminan las columnas “invoice_no” y la columna “quantity”, ya que no se utilizarán para los análisis posteriores.

In [ ]:
unificado_df.drop(['invoice_no', 'quantity'], axis=1, inplace=True)
unificado_df

customer_id         category    price     shopping_mall  gender   age  \
0         C241288         Clothing  1500.40            Kanyon  Female  28.0   
1         C111565            Shoes  1800.51    Forum Istanbul    Male  21.0   
2         C266599         Clothing   300.08         Metrocity    Male  20.0   
3         C988172            Shoes  3000.85      Metropol AVM  Female  66.0   
4         C189076            Books    60.60            Kanyon  Female  53.0   
...           ...              ...      ...               ...     ...   ...   
99452     C441542         Souvenir    58.65            Kanyon  Female  45.0   
99453     C569580  Food & Beverage    10.46    Forum Istanbul    Male  27.0   
99454     C103292  Food & Beverage    10.46         Metrocity    Male  63.0   
99455     C800631       Technology  4200.00      Istinye Park    Male  56.0   
99456     C273973         Souvenir    35.19  Mall of Istanbul  Female  36.0   

      payment_method year_month rango_etario  
0        Credit Card    2022-08      25 a 35  
1         Debit Card    2021-12          >24  
2               Cash    2021-11          >24  
3        Credit Card    2021-05          <61  
4               Cash    2021-10      46 a 60  
...              ...        ...          ...  
99452    Credit Card    2022-09      36 a 45  
99453           Cash    2021-09      25 a 35  
99454     Debit Card    2021-03          <61  
99455           Cash    2021-03      46 a 60  
99456    Credit Card    2022-10      36 a 45  

[99338 rows x 9 columns]

**Conclusión:** Al borrar las 2 columnas mencionadas y solicitar visualizar el DataFrame, observamos que la cantidad de columnas se reduce a 9.

12. Para pasar a la etapa de **Carga (LOAD)**, generamos un nuevo DataFrame limpio, es decir, con todas las modificaciones/actualizaciones realizadas en la etapa de Transformación.

In [ ]:
nuevo_df = unificado_df.copy()
nuevo_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 99338 entries, 0 to 99456
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype   
---  ------          --------------  -----   
 0   customer_id     99338 non-null  object  
 1   category        99338 non-null  object  
 2   price           99338 non-null  float64 
 3   shopping_mall   99338 non-null  object  
 4   gender          99338 non-null  object  
 5   age             99338 non-null  float64 
 6   payment_method  99338 non-null  object  
 7   year_month      99338 non-null  object  
 8   rango_etario    99338 non-null  category
dtypes: category(1), float64(2), object(6)
memory usage: 8.9+ MB


**Conclusión:** Generamos un nuevo DataFrame realizando una copia de unificado_df y mostramos la información de las columnas.

13. Aplicamos restricciones de integridad para asegurar la validez de los datos.

In [ ]:
initial_count = nuevo_df.shape[0]
nuevo_df = nuevo_df[(nuevo_df['age'] >= 17) & (nuevo_df['age'] <= 80)]
final_count = nuevo_df.shape[0]
print (f"Filas iniciales: {initial_count}, Filas después del filtro: {final_count}")

Filas iniciales: 99338, Filas después del filtro: 99338


**Conclusión:** En este caso, aplicamos una restricción a la columna "edad", donde solicitamos un conteo inicial de filas, luego aplicamos un filtro que guarde las edades entre 18 y 80 años, y que realice un conteo final. Al imprimir los resultados, corroboramos que tiene la misma cantidad de filas.

**Proceso de ANALISIS DE DATOS.**

A. Cantidades y Porcentaje de Mujeres y Hombres que contiene el DataFrame.

In [ ]:
porcentaje_genero = (nuevo_df['gender'].value_counts(normalize=True) * 100).round(2)
porcentaje = pd.DataFrame({'Porcentaje (%)': porcentaje_genero})
resultadoA_df = pd.merge(nuevo_df, porcentaje, on='gender', how='left')
resultadoA_df = resultadoA_df.groupby(['gender', 'Porcentaje (%)']).size().reset_index(name='count')
resultadoA_df

gender  Porcentaje (%)  count
0  Female           59.81  59412
1    Male           40.19  39926

**Conclusión:**
- línea1: realiza un conteo de la columna "gender" y lo transformamos a porcentaje con 2 decimales.
- línea2: transformamos los porcentajes en un dataframe
- línea3: combinamos los DataFrames con referencia a la columna "gender". (how='left') asegura que se mantengan los datos de nuevo_df y que los porcentajes se añadan a cada fila correspondiente.
- línea4: agrupa y convierte el DataFrame en columnas: "gender", Porcentaje" y "count".

B. Porcentaje de mujeres y hombres por rango etario

In [ ]:
genero_rango_df = nuevo_df.groupby(['gender', 'rango_etario']).size().reset_index(name='Cantidad')
genero_rango_df

<ipython-input-20-0aa574158d85>:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  genero_rango_df = nuevo_df.groupby(['gender', 'rango_etario']).size().reset_index(name='Cantidad')


gender rango_etario  Cantidad
0  Female          >24      7931
1  Female      25 a 35     12535
2  Female      36 a 45     11746
3  Female      46 a 60     16959
4  Female          <61     10241
5    Male          >24      5553
6    Male      25 a 35      8358
7    Male      36 a 45      7668
8    Male      46 a 60     11441
9    Male          <61      6906

C. Cantidad de consumidores por edad

In [ ]:
genero_edad_df = nuevo_df.groupby(['gender', 'age']).size().reset_index(name='Cantidad')
genero_edad_df

gender   age  Cantidad
0    Female  18.0      1078
1    Female  19.0      1130
2    Female  20.0      1102
3    Female  21.0      1165
4    Female  22.0      1207
..      ...   ...       ...
99     Male  65.0       737
100    Male  66.0       784
101    Male  67.0       766
102    Male  68.0       741
103    Male  69.0       757

[104 rows x 3 columns]

D. Modo de pago más frecuente de todos los clientes categorizados por su género.

In [ ]:
pago_genero_df = nuevo_df.groupby(['gender', 'payment_method']).size().reset_index(name='Cantidad')
pago_genero_df = pago_genero_df.sort_values(by='Cantidad', ascending=False).reset_index(drop=True)
pago_genero_df

gender payment_method  Cantidad
0  Female           Cash     26479
1  Female    Credit Card     20994
2    Male           Cash     17918
3    Male    Credit Card     13904
4  Female     Debit Card     11939
5    Male     Debit Card      8104

**Conclusión:** Podemos observar los datos organizados por género (femenino y masculino) organizados desde el método de pago más utilizado al menos utilizado.

E. Queremos determinar el método de pago más utilizado por el rango etario.

In [ ]:
pago_rango_df = nuevo_df.groupby(['payment_method', 'rango_etario']).size().reset_index(name='Cantidad')
pago_rango_df = pago_rango_df.sort_values(by='Cantidad', ascending=False).reset_index(drop=True)
pago_rango_df

<ipython-input-37-38ff27e948cf>:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  pago_rango_df = nuevo_df.groupby(['payment_method', 'rango_etario']).size().reset_index(name='Cantidad')


payment_method rango_etario  Cantidad
0            Cash      46 a 60     12626
1     Credit Card      46 a 60      9971
2            Cash      25 a 35      9356
3            Cash      36 a 45      8669
4            Cash          <61      7755
5     Credit Card      25 a 35      7274
6     Credit Card      36 a 45      6824
7     Credit Card          <61      6037
8            Cash          >24      5991
9      Debit Card      46 a 60      5803
10    Credit Card          >24      4792
11     Debit Card      25 a 35      4263
12     Debit Card      36 a 45      3921
13     Debit Card          <61      3355
14     Debit Card          >24      2701

F. Queremos determinar los precios de los productos por categoría

In [ ]:
precio_cat_df = nuevo_df.groupby(['category', 'price']).size().reset_index(name='Cantidad')
precio_cat_df

category    price  Cantidad
0             Books    15.15       984
1             Books    30.30      1007
2             Books    45.45      1000
3             Books    60.60       958
4             Books    75.75      1027
5          Clothing   300.08      6834
6          Clothing   600.16      6939
7          Clothing   900.24      6899
8          Clothing  1200.32      6838
9          Clothing  1500.40      6935
10        Cosmetics    40.66      2983
11        Cosmetics    81.32      2971
12        Cosmetics   121.98      3069
13        Cosmetics   162.64      3010
14        Cosmetics   203.30      3051
15  Food & Beverage     5.23      2999
16  Food & Beverage    10.46      2906
17  Food & Beverage    15.69      2990
18  Food & Beverage    20.92      2879
19  Food & Beverage    26.15      2987
20            Shoes   600.17      1941
21            Shoes  1200.34      2041
22            Shoes  1800.51      2006
23            Shoes  2400.68      2018
24            Shoes  3000.85      2009
25         Souvenir    11.73      1022
26         Souvenir    23.46       984
27         Souvenir    35.19      1076
28         Souvenir    46.92       920
29         Souvenir    58.65       989
30       Technology  1050.00       985
31       Technology  2100.00       999
32       Technology  3150.00      1009
33       Technology  4200.00      1001
34       Technology  5250.00      1001
35             Toys    35.84      1993
36             Toys    71.68      1958
37             Toys   107.52      2079
38             Toys   143.36      2072
39             Toys   179.20      1969

G. Solicitamos que nos muestre el precio mínimo, máximo y un promedio por categoría.

In [ ]:
precios_por_categoria_df = nuevo_df.groupby('category')['price'].agg(['min', 'max', 'mean']).reset_index()
precios_por_categoria_df.rename(columns={'min': 'precio_minimo', 'max': 'precio_maximo', 'mean': 'precio_promedio'}, inplace=True)
precios_por_categoria_df

category  precio_minimo  precio_maximo  precio_promedio
0            Books          15.15          75.75        45.562651
1         Clothing         300.08        1500.40       901.119898
2        Cosmetics          40.66         203.30       122.451725
3  Food & Beverage           5.23          26.15        15.671930
4            Shoes         600.17        3000.85      1807.281763
5         Souvenir          11.73          58.65        34.884470
6       Technology        1050.00        5250.00      3157.147147
7             Toys          35.84         179.20       107.754876

H. Solicitamos visualizar las categorías de los productos comprados por género y método de pago.

In [ ]:
categoria_genero_df = nuevo_df.groupby(['category', 'gender', 'payment_method']).size().reset_index(name='Cantidad')
categoria_genero_df = categoria_genero_df.sort_values(by='Cantidad', ascending=False).reset_index(drop=True)
categoria_genero_df

category  gender payment_method  Cantidad
0          Clothing  Female           Cash      9266
1          Clothing  Female    Credit Card      7175
2          Clothing    Male           Cash      6170
3          Clothing    Male    Credit Card      4840
4          Clothing  Female     Debit Card      4190
5   Food & Beverage  Female           Cash      3911
6         Cosmetics  Female           Cash      3903
7         Cosmetics  Female    Credit Card      3281
8   Food & Beverage  Female    Credit Card      3127
9          Clothing    Male     Debit Card      2804
10        Cosmetics    Male           Cash      2766
11             Toys  Female           Cash      2729
12  Food & Beverage    Male           Cash      2671
13            Shoes  Female           Cash      2652
14             Toys  Female    Credit Card      2154
15            Shoes  Female    Credit Card      2136
16  Food & Beverage    Male    Credit Card      2119
17        Cosmetics    Male    Credit Card      2052
18        Cosmetics  Female     Debit Card      1879
19            Shoes    Male           Cash      1813
20             Toys    Male           Cash      1806
21  Food & Beverage  Female     Debit Card      1754
22            Shoes    Male    Credit Card      1411
23             Toys    Male    Credit Card      1391
24       Technology  Female           Cash      1361
25         Souvenir  Female           Cash      1340
26            Books  Female           Cash      1317
27        Cosmetics    Male     Debit Card      1203
28             Toys  Female     Debit Card      1195
29  Food & Beverage    Male     Debit Card      1179
30            Shoes  Female     Debit Card      1167
31         Souvenir  Female    Credit Card      1077
32       Technology  Female    Credit Card      1035
33            Books  Female    Credit Card      1009
34            Books    Male           Cash       949
35       Technology    Male           Cash       873
36         Souvenir    Male           Cash       870
37            Shoes    Male     Debit Card       836
38             Toys    Male     Debit Card       796
39       Technology    Male    Credit Card       718
40         Souvenir    Male    Credit Card       689
41            Books    Male    Credit Card       684
42         Souvenir  Female     Debit Card       593
43       Technology  Female     Debit Card       585
44            Books  Female     Debit Card       576
45            Books    Male     Debit Card       441
46       Technology    Male     Debit Card       423
47         Souvenir    Male     Debit Card       422

I. Solicitamos visualizar la cantidad de compras por categoría según la fecha (año/mes).

In [ ]:
fecha_categoria_df = nuevo_df.groupby(['category', 'year_month']).size().reset_index(name='Cantidad')
fecha_categoria_df

category year_month  Cantidad
0      Books    2021-01       200
1      Books    2021-02       165
2      Books    2021-03       201
3      Books    2021-04       190
4      Books    2021-05       182
..       ...        ...       ...
211     Toys    2022-11       421
212     Toys    2022-12       396
213     Toys    2023-01       383
214     Toys    2023-02       368
215     Toys    2023-03        81

[216 rows x 3 columns]

J. Solicitamos visualizar el método de pago acorde al precio.

In [ ]:
pago_precio_df = nuevo_df.groupby(['price', 'payment_method']).size().reset_index(name='Cantidad')
pago_precio_df = pago_precio_df.sort_values(by='Cantidad', ascending=False).reset_index(drop=True)
pd.set_option("display.max_rows", None)

# Mostrar todas las columnas del DataFrame
pd.set_option("display.max_columns", None)
pago_precio_df

price payment_method  Cantidad
0    1500.40           Cash      3146
1     600.16           Cash      3123
2     900.24           Cash      3117
3    1200.32           Cash      3036
4     300.08           Cash      3014
5     300.08    Credit Card      2467
6    1200.32    Credit Card      2425
7     600.16    Credit Card      2392
8     900.24    Credit Card      2381
9    1500.40    Credit Card      2350
10   1500.40     Debit Card      1439
11    600.16     Debit Card      1424
12    900.24     Debit Card      1401
13   1200.32     Debit Card      1377
14      5.23           Cash      1376
15     40.66           Cash      1363
16    300.08     Debit Card      1353
17    121.98           Cash      1341
18    162.64           Cash      1340
19     26.15           Cash      1325
20     81.32           Cash      1320
21     15.69           Cash      1309
22     10.46           Cash      1306
23    203.30           Cash      1305
24     20.92           Cash      1266
25    203.30    Credit Card      1127
26     15.69    Credit Card      1098
27    162.64    Credit Card      1087
28    121.98    Credit Card      1072
29     26.15    Credit Card      1054
30      5.23    Credit Card      1035
31     20.92    Credit Card      1033
32     81.32    Credit Card      1031
33     10.46    Credit Card      1026
34     40.66    Credit Card      1016
35    143.36           Cash       951
36    107.52           Cash       936
37   2400.68           Cash       918
38   3000.85           Cash       900
39    179.20           Cash       897
40   1200.34           Cash       892
41     71.68           Cash       878
42   1800.51           Cash       878
43    600.17           Cash       877
44     35.84           Cash       873
45    107.52    Credit Card       738
46   1800.51    Credit Card       737
47   1200.34    Credit Card       728
48    143.36    Credit Card       728
49     35.84    Credit Card       715
50   2400.68    Credit Card       715
51   3000.85    Credit Card       701
52    179.20    Credit Card       698
53    600.17    Credit Card       666
54     71.68    Credit Card       666
55    121.98     Debit Card       656
56     81.32     Debit Card       620
57    203.30     Debit Card       619
58     26.15     Debit Card       608
59     40.66     Debit Card       604
60      5.23     Debit Card       588
61     15.69     Debit Card       583
62    162.64     Debit Card       583
63     20.92     Debit Card       580
64     10.46     Debit Card       574
65     35.19           Cash       468
66     30.30           Cash       468
67     11.73           Cash       468
68     75.75           Cash       463
69     45.45           Cash       460
70   5250.00           Cash       452
71   4200.00           Cash       450
72   1050.00           Cash       450
73     23.46           Cash       449
74   3150.00           Cash       444
75     60.60           Cash       440
76   2100.00           Cash       438
77     15.15           Cash       435
78   1200.34     Debit Card       421
79     71.68     Debit Card       414
80     46.92           Cash       414
81     58.65           Cash       411
82   3000.85     Debit Card       408
83    107.52     Debit Card       405
84     35.84     Debit Card       405
85    600.17     Debit Card       398
86     35.19    Credit Card       397
87    143.36     Debit Card       393
88   1800.51     Debit Card       391
89   2400.68     Debit Card       385
90   3150.00    Credit Card       385
91    179.20     Debit Card       374
92     58.65    Credit Card       369
93     11.73    Credit Card       356
94   4200.00    Credit Card       353
95   5250.00    Credit Card       348
96     30.30    Credit Card       347
97   2100.00    Credit Card       346
98     15.15    Credit Card       343
99     75.75    Credit Card       341
100    45.45    Credit Card       340
101    23.46    Credit Card       332
102    60.60    Credit Card       322
103  1050.00    Credit Card       321
104    46.92    C

K. Solicitamos visualizar el lugar de compra (shopping_mall) acorde al género y rango etario.

In [ ]:
genero_lugar_df = nuevo_df.groupby(['gender', 'shopping_mall', 'rango_etario']).size().reset_index(name='Cantidad')
genero_lugar_df = genero_lugar_df.sort_values(by='Cantidad', ascending=False).reset_index(drop=True)
pd.set_option("display.max_rows", None)

# Mostrar todas las columnas del DataFrame
pd.set_option("display.max_columns", None)
genero_lugar_df

<ipython-input-37-0876272f2c9f>:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  genero_lugar_df = nuevo_df.groupby(['gender', 'shopping_mall', 'rango_etario']).size().reset_index(name='Cantidad')


gender      shopping_mall rango_etario  Cantidad
0   Female             Kanyon      46 a 60      3399
1   Female   Mall of Istanbul      46 a 60      3389
2   Female          Metrocity      46 a 60      2526
3   Female   Mall of Istanbul      25 a 35      2463
4   Female             Kanyon      25 a 35      2459
5   Female             Kanyon      36 a 45      2375
6   Female   Mall of Istanbul      36 a 45      2319
7     Male             Kanyon      46 a 60      2306
8     Male   Mall of Istanbul      46 a 60      2294
9   Female   Mall of Istanbul          <61      2118
10  Female             Kanyon          <61      2050
11  Female          Metrocity      25 a 35      1900
12  Female          Metrocity      36 a 45      1779
13    Male          Metrocity      46 a 60      1768
14  Female       Metropol AVM      46 a 60      1712
15    Male   Mall of Istanbul      25 a 35      1695
16  Female       Istinye Park      46 a 60      1660
17  Female             Kanyon          >24      1611
18    Male             Kanyon      25 a 35      1598
19  Female   Mall of Istanbul          >24      1598
20    Male             Kanyon      36 a 45      1556
21    Male   Mall of Istanbul      36 a 45      1550
22  Female          Metrocity          <61      1533
23    Male   Mall of Istanbul          <61      1358
24    Male             Kanyon          <61      1354
25  Female       Metropol AVM      25 a 35      1340
26    Male          Metrocity      25 a 35      1279
27  Female       Metropol AVM      36 a 45      1260
28  Female       Istinye Park      25 a 35      1241
29  Female          Metrocity          >24      1194
30    Male          Metrocity      36 a 45      1142
31  Female       Istinye Park      36 a 45      1141
32    Male       Metropol AVM      46 a 60      1140
33    Male   Mall of Istanbul          >24      1130
34    Male       Istinye Park      46 a 60      1110
35    Male             Kanyon          >24      1097
36    Male          Metrocity          <61      1068
37  Female       Istinye Park          <61      1046
38  Female       Metropol AVM          <61       993
39  Female       Zorlu Center      46 a 60       891
40  Female     Forum Istanbul      46 a 60       867
41  Female        Cevahir AVM      46 a 60       858
42    Male       Istinye Park      25 a 35       856
43  Female     Viaport Outlet      46 a 60       845
44  Female       Metropol AVM          >24       828
45  Female  Emaar Square Mall      46 a 60       812
46    Male          Metrocity          >24       807
47    Male       Metropol AVM      25 a 35       803
48  Female       Istinye Park          >24       777
49    Male       Metropol AVM      36 a 45       772
50    Male       Istinye Park      36 a 45       745
51    Male       Metropol AVM          <61       725
52  Female     Forum Istanbul      25 a 35       669
53  Female        Cevahir AVM      25 a 35       662
54    Male       Istinye Park          <61       653
55    Male       Zorlu Center      46 a 60       622
56  Female     Viaport Outlet      36 a 45       609
57  Female       Zorlu Center      25 a 35       603
58  Female     Viaport Outlet      25 a 35       602
59  Female  Emaar Square Mall      25 a 35       596
60  Female       Zorlu Center      36 a 45       585
61    Male       Metropol AVM          >24       573
62  Female  Emaar Square Mall      36 a 45       566
63  Female        Cevahir AVM      36 a 45       564
64    Male        Cevahir AVM      46 a 60       557
65    Male  Emaar Square Mall      46 a 60       556
66    Male     Viaport Outlet      46 a 60       548
67  Female     Forum Istanbul      36 a 45       548
68    Male     Forum Istanbul      46 a 60       540
69    Male       Istinye Park          >24       536
70  Female     Forum Istanbul          <61       529
71  Female     Viaport Outlet          <61       509
72  Female       Zorlu Center          <61       500
73  Female  Emaar Square Mall          <61       496
74  Female        Cevahir AVM

L. Solicitamos visualizar el método de pago acorde al lugar de compra (shopping_mall).

In [ ]:
pago_lugar_df = nuevo_df.groupby(['payment_method', 'shopping_mall']).size().reset_index(name='Cantidad')
pago_lugar_df = pago_lugar_df.sort_values(by='Cantidad', ascending=False).reset_index(drop=True)
pago_lugar_df

payment_method      shopping_mall  Cantidad
0            Cash   Mall of Istanbul      8882
1            Cash             Kanyon      8845
2     Credit Card   Mall of Istanbul      7012
3     Credit Card             Kanyon      6910
4            Cash          Metrocity      6619
5     Credit Card          Metrocity      5342
6            Cash       Metropol AVM      4553
7            Cash       Istinye Park      4430
8      Debit Card             Kanyon      4050
9      Debit Card   Mall of Istanbul      4020
10    Credit Card       Metropol AVM      3519
11    Credit Card       Istinye Park      3415
12     Debit Card          Metrocity      3035
13           Cash       Zorlu Center      2320
14           Cash        Cevahir AVM      2228
15           Cash     Viaport Outlet      2228
16           Cash     Forum Istanbul      2181
17           Cash  Emaar Square Mall      2111
18     Debit Card       Metropol AVM      2074
19     Debit Card       Istinye Park      1920
20    Credit Card        Cevahir AVM      1779
21    Credit Card       Zorlu Center      1757
22    Credit Card     Forum Istanbul      1748
23    Credit Card     Viaport Outlet      1721
24    Credit Card  Emaar Square Mall      1695
25     Debit Card     Forum Istanbul      1014
26     Debit Card  Emaar Square Mall       999
27     Debit Card       Zorlu Center       988
28     Debit Card        Cevahir AVM       982
29     Debit Card     Viaport Outlet       961

M. Solicitamos visualizar los precios de compra acorde al género de la persona.

In [ ]:
genero_precio_df = nuevo_df.groupby(['gender', 'price']).size().reset_index(name='Cantidad')
genero_precio_df

gender    price  Cantidad
0   Female     5.23      1803
1   Female    10.46      1700
2   Female    11.73       604
3   Female    15.15       567
4   Female    15.69      1780
..     ...      ...       ...
75    Male  2400.68       806
76    Male  3000.85       820
77    Male  3150.00       435
78    Male  4200.00       402
79    Male  5250.00       392

[80 rows x 3 columns]

N. Solicitamos visualizar las categorías de compra acorde a la fecha (año/mes).

In [ ]:
nuevo_df['year_month'] = pd.to_datetime(nuevo_df['year_month'], format='%Y-%m')
nuevo_df['Year'] = nuevo_df['year_month'].dt.year
nuevo_df['Month'] = nuevo_df['year_month'].dt.month
categoria_month_year_df = nuevo_df.groupby(['Month', 'Year', 'category']).size().reset_index(name='Cantidad')
categoria_month_year_df = categoria_month_year_df.sort_values(by=['Month', 'Year']).reset_index(drop=True)
categoria_month_year_df

Month  Year         category  Cantidad
0        1  2021            Books       200
1        1  2021         Clothing      1327
2        1  2021        Cosmetics       555
3        1  2021  Food & Beverage       557
4        1  2021            Shoes       415
..     ...   ...              ...       ...
211     12  2022  Food & Beverage       568
212     12  2022            Shoes       402
213     12  2022         Souvenir       175
214     12  2022       Technology       198
215     12  2022             Toys       396

[216 rows x 4 columns]

ANALISIS DE CATEGORIAS POR FECHAS

In [ ]:
categoria_month_year_df

Month  Year         category  Cantidad
0        1  2021            Books       200
1        1  2021         Clothing      1327
2        1  2021        Cosmetics       555
3        1  2021  Food & Beverage       557
4        1  2021            Shoes       415
..     ...   ...              ...       ...
211     12  2022  Food & Beverage       568
212     12  2022            Shoes       402
213     12  2022         Souvenir       175
214     12  2022       Technology       198
215     12  2022             Toys       396

[216 rows x 4 columns]

Visualizamos las cantidades por categoría en orden descendente (para obtener un ranking)

In [ ]:
ranking_category_df = categoria_month_year_df.groupby('category')['Cantidad'].sum().reset_index()
ranking_category_df = ranking_category_df.sort_values(by='Cantidad', ascending=False).reset_index(drop=True)
ranking_category_df

category  Cantidad
0         Clothing     34445
1        Cosmetics     15084
2  Food & Beverage     14761
3             Toys     10071
4            Shoes     10015
5       Technology      4995
6         Souvenir      4991
7            Books      4976

Aplicamos un filtro de datos a visualizar sobre la categoría que nos interesa, en este caso, “Clothing” que es la primera.

In [ ]:
clothing_2021_df = categoria_month_year_df[(categoria_month_year_df['category'] == 'Clothing') & (categoria_month_year_df['Year'] == 2021)]
clothing_2021_df = clothing_2021_df.sort_values(by='Month', ascending=True).reset_index(drop=True)
clothing_2021_df

Month  Year  category  Cantidad
0       1  2021  Clothing      1327
1       2  2021  Clothing      1187
2       3  2021  Clothing      1374
3       4  2021  Clothing      1364
4       5  2021  Clothing      1320
5       6  2021  Clothing      1282
6       7  2021  Clothing      1470
7       8  2021  Clothing      1233
8       9  2021  Clothing      1291
9      10  2021  Clothing      1334
10     11  2021  Clothing      1404
11     12  2021  Clothing      1371

In [ ]:
clothing_2022_df = categoria_month_year_df[(categoria_month_year_df['category'] == 'Clothing') & (categoria_month_year_df['Year'] == 2022)]
clothing_2022_df = clothing_2022_df.sort_values(by='Month', ascending=True).reset_index(drop=True)
clothing_2022_df

Month  Year  category  Cantidad
0       1  2022  Clothing      1277
1       2  2022  Clothing      1219
2       3  2022  Clothing      1348
3       4  2022  Clothing      1275
4       5  2022  Clothing      1301
5       6  2022  Clothing      1299
6       7  2022  Clothing      1377
7       8  2022  Clothing      1306
8       9  2022  Clothing      1265
9      10  2022  Clothing      1339
10     11  2022  Clothing      1241
11     12  2022  Clothing      1289

In [ ]:
clothing_2023_df = categoria_month_year_df[(categoria_month_year_df['category'] == 'Clothing') & (categoria_month_year_df['Year'] == 2023)]
clothing_2023_df = clothing_2021_df.sort_values(by='Month', ascending=True).reset_index(drop=True)
clothing_2023_df

Month  Year  category  Cantidad
0       1  2021  Clothing      1327
1       2  2021  Clothing      1187
2       3  2021  Clothing      1374
3       4  2021  Clothing      1364
4       5  2021  Clothing      1320
5       6  2021  Clothing      1282
6       7  2021  Clothing      1470
7       8  2021  Clothing      1233
8       9  2021  Clothing      1291
9      10  2021  Clothing      1334
10     11  2021  Clothing      1404
11     12  2021  Clothing      1371